### PR TITLE
Update all batches on select purpose form

### PR DESCRIPTION
**Pull request by constanzauanini**

### CODE REVIEW

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 1]
```diff
@@ -72,33 +67,41 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
-            },
-          ],
-        });
-      } else {
-        snackbar.toastError(response.error);
-      }
+    const fetchBatches = () => {
+      const withdrawalBatches: NurseryWithdrawal[] = [];
+      batchIds.forEach(async (batchId) => {
+        const response = await getBatch(Number(batchId));
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:82]: This will be very slow. Best to do one search API with a values field that is a list of all batch ids.
Should this come from the parent component  as a prop `originalBatches` or here is ok? if other components need batches then should come from parent. I think the grid of selecting quantities form needs batches (filtered down by selected nursery).
We may need to maintain a local state of records that is a 1:1 mapping from original batch to a record.
every time we update this local record, we can update the parent record based on a filtered down list of records (based on from nursery).

Basically:
```
Props {
  batches: any [] // not typed since it comes from search
  records;
  setRecords;
};

const [localRecords, setLocalRecords] = useState<>([]);
useEffect() {
   populate localREcords with NurseryWithdrawal
}

// for every change in the form
onChangeXYZ( {
   setLocalRecords(new changes)
   if (fromNursery) { // set records in parent, filtered by from nursery
      setRecords(localRecords.filter(facilityId === fromNursery)
   }
}
```

File: src/components/Inventory/withdraw/BatchWithdrawFlow.tsx [hunk 2]
```diff
@@ -11,15 +15,34 @@ type BatchWithdrawFlowProps = {
 
 export default function BatchWithdrawFlow(props: BatchWithdrawFlowProps): JSX.Element {
   const { organization, batchIds } = props;
-  const [, setFlowState] = useState<FlowStates>('purpose' as unknown as FlowStates);
+  const [flowState, setFlowState] = useState<FlowStates>('purpose');
+  const [record, setRecord] = useForm<NurseryWithdrawal[]>([]);
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/BatchWithdrawFlow.tsx:19]: plural records and setRecords?

**Review by karthikbtf — COMMENTED** [review]
This is really awesome!! The approach is great.
Some comments which I think need follow up.

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 3]
```diff
@@ -176,71 +170,81 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
                     label={strings.FROM_NURSERY}
-                    selectedValue={batch.facilityId.toString()}
-                    options={[getNurseriesById(organization, batch.facilityId)].map((nursery) => ({
+                    selectedValue={record[0].facilityId.toString()}
+                    options={[getNurseriesById(organization, record[0].facilityId)].map((nursery) => ({
                       label: nursery.name,
                       value: nursery.id.toString(),
                     }))}
-                    onChange={(value) => onChange('destinationFacilityId', value)}
+                    onChange={() => true}
```
[the remark this answers is not in the document]
Comment (karthikbtf) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:213]: ignore this comment ^, my earlier comment addresses it with `onChangeXYZ stuff`

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 4]
```diff
@@ -176,71 +170,81 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
+              </Grid>
+
+              <Grid display='flex'>
                 <Grid item xs={isNurseryTransfer ? 6 : 12} sx={{ marginTop: theme.spacing(2) }} paddingRight={1}>
                   <Dropdown
                     id='from'
                     label={strings.FROM_NURSERY}
-                    selectedValue={batch.facilityId.toString()}
-                    options={[getNurseriesById(organization, batch.facilityId)].map((nursery) => ({
+                    selectedValue={record[0].facilityId.toString()}
```
[the remark this answers is not in the document]
Comment (constanzauanini) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx]: But aren't they all from the same nursery? If not we don't show the withdraw button

[the remark this answers is not in the document]
Comment (karthikbtf) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx]: If they come from inventory they won’t all be from same nursery

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 5]
```diff
@@ -176,71 +182,81 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
-            </Grid>
-            <Grid item xs={12} sx={{ marginTop: theme.spacing(2) }}>
-              <Textfield id='notes' value={record.notes} onChange={onChange} type='textarea' label={strings.NOTES} />
+                {isNurseryTransfer && (
+                  <Grid item xs={6} sx={{ marginTop: theme.spacing(2) }} paddingLeft={1}>
+                    <Dropdown
+                      id='destinationFacilityId'
+                      label={strings.TO_NURSERY_REQUIRED}
+                      selectedValue={firstBatch.destinationFacilityId?.toString()}
+                      options={getAllNurseries(organization).map((nursery) => ({
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:224]: Missed this. The list should only be nurseries that the batches belong to.
This feature is primarily for narrowing down nursery for batches coming from inventory page.

Comment (karthikbtf) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:224]: For batches coming from details page this will end up being a list of one nursery

Comment (karthikbtf) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:224]: Hence auto selected / defaulted to that value

File: src/components/Inventory/withdraw/BatchWithdrawFlow.tsx [hunk 6]
```diff
@@ -32,13 +75,13 @@ export default function BatchWithdrawFlow(props: BatchWithdrawFlowProps): JSX.El
 
   return (
     <>
-      {flowState === 'purpose' && (
+      {flowState === 'purpose' && batches && (
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/BatchWithdrawFlow.tsx:78]: i think we can move the batches check above?

```
if (!batches) {
  return <CircularProgress />
}
```
so we can avoid the batches check on all the components below

File: src/components/Inventory/withdraw/BatchWithdrawFlow.tsx [hunk 7]
```diff
@@ -16,7 +17,49 @@ type BatchWithdrawFlowProps = {
 export default function BatchWithdrawFlow(props: BatchWithdrawFlowProps): JSX.Element {
   const { organization, batchIds } = props;
   const [flowState, setFlowState] = useState<FlowStates>('purpose');
-  const [record, setRecord] = useForm<NurseryWithdrawal[]>([]);
+  const [records, setRecords] = useForm<NurseryWithdrawal[]>([]);
+  const [batches, setBatches] = useState<any[]>();
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/BatchWithdrawFlow.tsx:21]: 👍 👍

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 8]
```diff
@@ -78,41 +70,28 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
   useEffect(() => {
     const fetchBatches = () => {
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:71]: `populateWithdrawals`?

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 9]
```diff
@@ -146,25 +125,55 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
+    if (localRecords) {
       const nurseryTransferInvalid = !validateNurseryTransfer();
-      if (fieldsErrors.withdrawDate || nurseryTransferInvalid) {
+      const selectedNurseryInvalid = !validateSelectedNursery();
+      if (fieldsErrors.withdrawDate || nurseryTransferInvalid || selectedNurseryInvalid) {
         return;
       }
     }
 
+    setRecords(localRecords.filter((rcd) => rcd.facilityId.toString() === selectedNursery));
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:153]: nice, this is a good place to do this

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 10]
```diff
@@ -146,25 +125,55 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
 
+  const validateSelectedNursery = () => {
+    if (!selectedNursery) {
+      setIndividualError('fromFacilityId', strings.REQUIRED_FIELD);
+      return false;
+    }
+    return true;
+  };
+
   const saveWithdrawal = async () => {
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:144]: `onNext`?

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 11]
```diff
@@ -146,25 +125,55 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
     }
 
+    setRecords(localRecords.filter((rcd) => rcd.facilityId.toString() === selectedNursery));
+
     onNext();
   };
 
+  const onChangeFromNursery = (facilityIdSelected: string) => {
+    setSelectedNursery(facilityIdSelected);
+    setRecords(records.filter((localR) => localR.facilityId.toString() === facilityIdSelected));
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:160]: we can remove this maybe, since you are handling this on line 153

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx (hunk 5 above)
Comment (constanzauanini) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:233]: But if we have selected only one batch and we want to do a "Nursery Transfer", would it be "from" the nursery it belongs "to" the nursery it belongs?

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 12]
```diff
@@ -146,25 +125,55 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
+  };
+
+  const getNurseriesOptions = () => {
+    const nurseriesOptions: DropdownItem[] = [];
+    batches.forEach((bat) => {
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:165]: Optional suggestion (since `find` is a sequential search, we can use assoc array). Current code is just as fine.
```
const nurseries = batches.reduce((acc, batch) => {
  if (!acc[batch.facility_id.toString()]) { // this lookup may be faster
     acc[batch.facility_id.toString()] = { label: batch.facility_id, value: batch.facility_name};
  }
  return acc;
}, {});
const options = Object.values(nurseries); // or Object.keys(nurseries).map(id => nurseries[id]);
if (options.length === 1) {
  setSelectedNursery(...)
}
return options;
```

**Review by karthikbtf — APPROVED** [review]
This is very nice, matches the requirements. Some suggestions for followup.

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx [hunk 13]
```diff
@@ -72,32 +68,30 @@ export default function SelectPurposeForm(props: SelectPurposeFormProps): JSX.El
...
-      const response = await getBatch(Number(batchId));
-      if (response.requestSucceeded && response.batch) {
-        setBatch(response.batch);
-        setRecord({
+    const fetchBatches = () => {
+      const withdrawalBatches: NurseryWithdrawal[] = [];
+      batches.forEach((batch, index) => {
+        const withdrawalBatch: NurseryWithdrawal = {
+          id: -1,
           purpose: 'Out Plant',
```
Comment (karthikbtf) [inline src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:76]: assuming you will change this to something else and hide outplant for contributors

File: src/components/Inventory/withdraw/BatchWithdrawFlow.tsx (hunk 6 above)
Comment (constanzauanini) [reply src/components/Inventory/withdraw/BatchWithdrawFlow.tsx:78]: Good idea

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx (hunk 8 above)
Comment (constanzauanini) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:71]: Yes, sorry, this was the old name

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx (hunk 13 above)
Comment (constanzauanini) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:76]: Yes, will do it in separate PR

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx (hunk 5 above)
Comment (karthikbtf) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:234]: I ended up commenting on the wrong code, I was referring to the list of 'from nurseries' - which you have already fixed. my bad!!

Comment (karthikbtf) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:234]: btw do we may need to filter out the from nursery from this list? we can do that as a follow up

**Review by karthikbtf — APPROVED** [review]
🎉 🎉 🎉  looks great, thanks @constanzauanini

File: src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx (hunk 5 above)
Comment (constanzauanini) [reply src/components/Inventory/withdraw/flow/SelectPurposeForm.tsx:234]: Correct, I will do this in a following PR
